### ISSUE TITLE
Update helloworld-tutorial to include protos split over multiple files

### ISSUE THREAD

**Issue by alelazar** [hyperium/tonic#1003]
## Feature Request

I believe the tutorial at https://github.com/hyperium/tonic/blob/master/examples/helloworld-tutorial.md should be updated to put the `HelloReply` and `HelloRequest` in a separate file from the service

### Motivation

Despite a lot of experience with proto & gRPC in golang & python, I had trouble making the basic switch to split files

Tonic & Rust present some style differences that are not present in other languages. Protos in multiple files is a very simple thing that should have better examples. It is also not directly accessible via documentation. 

### Alternatives

Give more explicit documentation on how to handle splitting/importing protos across files

### Changes Required

I will put up a PR for this shortly

**Comment by LucioFranco** [comment]
Yeah this seems like something we could show at the end as an extra maybe?

**Comment by EdmundsEcho** [comment]
It would be great.  Having done an hour or two of reading various specs, I'm beginning to think the precise manner by which `tonic_build` works with multiple files is an "implementation-specific" strategy.  In other words, a strategy where the google spec ends.  The important content would just matter of fact (i.e., no explanation required) present:

1. possible to use multiple .proto files to specify a single service? Yes.
2. possible to use multiple .proto files to specify multiple services? Yes, but likely bad practice b/c breaks ability to deliver on "single wire" benefit
3. possible to reference messages from other .proto? Yes 
    * How so when the message spec is within the project_root/proto/
    * How so when the message spec is outside the project

For all of the above, 
1. how we import and reference the messages in the proto spec
2. how we specify the location in the build script
3. how that translates to auto-generated rust types

For instance, based on my experience with graphQl, once I specify the root graphql, the parser uses the equivalent of the .proto imports to "figure it out".  So, I would expect (as a "where I'm starting from") the build script to require only the root.  

However, Prost may have gone another way given for instance Rust allows us to explicitly express the map between modules and the path to where it is defined.  I suspect in this case it gets a bit more complicated given that we already have a source of truth specified by how the proto files already have a import facility that dictates a would be "competing/alternative truth" (if you will :)).

**Comment by EdmundsEcho** [comment]
Suggested copy for the tutorial.

Scope: `proto -> rust`

## With the end in mind

```rs
/// run tokio main
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let addr = "[::1]:50051".parse()?;

    // dependency injection; pass-in an instantiated instance of our proto service
 
    // we inject...
    // instance of our service that implements the auto-generated proto service trait
    let root_service = RootService::default();

    // into `Server`, a concrete type specified by tonic
    // in `tonic::transport`; 
    // 
    // My service implements the auto-generated trait found as a child of the 
    // root module: root::root_server.
    //  
    // There is a wrapper between the `Server` and "my service": `RootServer`.
    // The wrapper is an auto-generated concrete type found as a child of the
    // root module root::root_server already described.
 
    Server::builder()
        .add_service(RootServer::new(root_service))
        .serve(addr)
        .await?;

    Ok(())
}
```

## The implementation of what gets injected

```rs
// something that implements the auto-generated trait
// ... wrapped by a concrete type that is also auto-generated
```

It implements a auto-generated trait that encodes the  service and `rpc`s and `Messages`.

|rust|proto|Comment|
|:---|:---|:---|
|trait|service| only one for a given "wire"|
|trait method|rpc||
|types|Message| primitives and compound types (messages specified by other messages|
|mod|package| some implementations of an enumerated type also uses mod|

## Key to specifying messages specified in multiple files

### The proto files
```fish
/proto/root.proto
/proto/root.shared.proto
```

Where root.proto is tagged with `package root`, and root.shared is tagged with `package root.shared`.  The `root.proto` `imports root.shared.proto` and references a shared message using `root.SharedMessage`.

1. Be aware that the latest `prost` will translate nested package names to a nested package (e.g., in the *root.rs*:  `package root.shared` ->  `super::shared` (evidence of the nested package structure isn't apparent in the *root.shared.rs*).

2. In the `server.rs` file declare the nested modules using the `tonic::include_proto!` macros:

```rs
// server.rs
pub mod root {
    tonic::include_proto!("root"); // proto package root
    
    pub mod shared {
        tonic::include_proto!("root.shared"); // proto package root.shared
    }
}
```

3.  The declaration in (2) also provides what I need to reference the, now rust, proto messages and service trait:

```rs
// server.rs
// reference the `mod root_server` 
use root::root_server::Root; // trait
use root::root_server::RootServer; // a concrete type
 
// Do not reference the Messages specified in root.proto, 
// more generally, any Message in proto  
// "includes" (non-root) proto package, now in rust  
// use root::SomeRootMessage; -- this is wrong 

// we have all that is required to reference 
// the proto -> rust files

// trait implementation on struct RootService {}
impl Root for RootService...

// root Messages
let aRootMessage = root::ARootMessage...;
let aSharedMessage =
root::shared::ASharedMessage;
```

### Summary

A. The `root` proto triggers the code generation for 3 ingredients
1. Trait that specifies our service
2. The concrete type that consumes whatever we use to implement the service trait
3. The Message types specified in the root.proto

B. The "non-root" proto files (aka "includes"), triggers the generation for one thing: additional Messages. How to reference these Messages in a rust context:
1. packages -> packages that know how to reference each other
2. declare the mods, nested as-need-be to reflect the package declarations
3. use the `tonic::include_proto!("<package>");` to fill-out the body of each of the mod.

C. ... I used the `server.rs` to put all of the ingredients together.

**Comment by michalisioak** [comment]
added this example hope it helps #1384
